text/BrunoPitrus/galil.tex

```diff
@@ -0,0 +1,117 @@
+   s' < s + p2 \textit{takie że} word[s'+1:] \textit{nie ma k-okresu prefiksu krótszego niż} p2
+  s= s'
+\end{alltt}
+Niech \texttt{l(s)} będzie długością najkrótszego okresu \texttt{word[s+1:]} a \texttt{p1(s)} długością najkrótszego \(k\)-okresu prefiksu. Indukcyjnie można dowieść, że w każdej iteracji pętli \texttt{s < c * min(l(s), p1(s))} dla \texttt{c = (k-1)/(k-2)} — więc w szczególności na końcu \texttt{len(u) < c * l(s)}.
+
+W szczególności istnieje następujący prosty algorytm znajdowania \texttt{s'}:
+\begin{alltt}
+s'= s
+\textbf{while} word[s'+1:] \textit{ma k-okres prefiksu krótszy niż p2}:
+  usuń najkrótszy okres prefiksu
+\end{alltt}
+Przeanalizujmy złożoność ostatecznego kodu programu:
+\begin{alltt}
+K= 4
+s,p1,q1= 0,1,0
+p2,q2= 0,0
+mode=0
+\textbf{while} True:
+  \textbf{if} mode==1:
+    \textit{#Znajdź najkrótszy k-okres prefiksu word[s+1:]}
+    \textbf{while} s+p1+q1 < m \textbf{and} word[s+p1+q1+1] == word[s+q1+1]: q1+= 1
+    \textbf{if} p1+q1 >= p1*K: p2, q2= q1,0; mode=2; \textbf{continue}
+    \textbf{if} s+p1+q1 == m: \textbf{break}
+    p1+= (1 \textbf{if} q1==0 \textbf{else} (q1+K-1)//K); q1= 0
+  \textbf{elif} mode==2:
+    \textit{#Znajdź drugi najkrótszy k-okres prefiksu word[s+1:]. Jeśli nie istnieje to przejdź do drugiej fazy algorytmu.}
+    \textbf{while} s+p2+q2 < m \textbf{and} word[s+p2+q2+1] == word[s+q2+1] \textbf{and} p2+q2 < p2*K: q2+= 1
+    \textbf{if} p2+q2 == p2*K: mode= 0; \textbf{continue};
+    \textbf{if} s+p2+q2 == m: \textbf{break}
+    \textbf{if} q2 == p1+q1:
+      p2+= p1; q2-= p1;
+    \textbf{else}:
+      p2+= (1 \textbf{if} q2==0 \textbf{else} (q2+K-1)//K); q2= 0
+  \textbf{else}:
+    \textit{#Zinkrementuj s}
+    \textbf{while} s+p1+q1 < m \textbf{and} word[s+p1+q1+1] == word[s+q1+1]: q1+= 1
+    \textbf{while} p1+q1 >= p1*K: s+= p1; q1-= p1;
+    p1+= (1 \textbf{if} q1==0 \textbf{else} (q1+K-1)//K); q1= 0
+    \textbf{if} p1 >= p2: mode= 1
+\end{alltt}
+W celu przeanalizowania złożoności tej części algorytmu rozważmy wyrażenie
+\[
+2s+{\left(k+1\right)}p_1+q_1+{\left(k+1\right)}p_2+q_2
+\]
+Jego wartość jest zawsze \(\symrm O{\left(\texttt{len(word)}\right)}\) a każde przypisanie ją zwiększa.
+\begin{alltt}
+p2,q2= 0,0
+\textbf{while} True:
+  \textbf{while} p2+s+q2 < n \textbf{and} s+q2 < m \textbf{and} text[p2+s+q2+1] == word[s+q2+1]: q2+= 1
+  \textbf{if} q2 == m-s \textbf{and} text[p2+1 : p2+s+1] == word[1 : s+1]:
+    yield p2+1
+  \textbf{if} q2 == p1+q1:
+    p2+= p1; q2-= p1
+  \textbf{else}:
+    p2+= (1 \textbf{if} q2==0 \textbf{else} (q2+K-1)//K); q2= 0
+  \textbf{if} p2+s > n: \textbf{return}
+\end{alltt}
+Przed wykonaniem drugiej części algorytmu jeśli \texttt{word[s+1:]} ma \(k\)-okres prefiksu, to jego długość wynosi \(p_1\).
+Ponadto twierdzimy, że \(s<{\left(k-1\right)}p_1/{\left(k-2\right)}\) co nam zagwarantuje że bezpośrednie sprawdzanie równości podciągów zajmie łącznie \(\symrm O{\left(\texttt{len(text)}\right)}\) czasu.

```

Comment:
That's not obvious, but I guess it refers to some (unexplained) claim above. Please relate it.
Moreover, I don't see how "len(u) < c * l(s)" (which you claim above) is related to a linear-time execution of this algorithm.